core/src/main/java/com/yahoo/oak/NovaMMHeader.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2020, Verizon Media.
+ * Licensed under the terms of the Apache 2.0 license.
+ * Please see LICENSE file in the project root for terms.
+ */
+
+package com.yahoo.oak;
+
+import com.yahoo.oak.ValueUtils.ValueResult;
+
+import java.util.concurrent.atomic.AtomicLongArray;
+
+
+class NovaMMHeader {
+
+    /*
+    * Long NovaMMHeader: int version + int lock
+    * 0...  ...21 | 32...                  ...63
+    *  version    |  length
+    *
+    * The length (Mostly in integer size) of the data is also held just following the header (since the block and
+    * offset are 42 bits at most the length can be 42 bits which we allow here!)
+    * The length is set once upon header allocation and later can only be read. */
+
+    private static final int VERSION_SIZE = 22;  //block offset get 42 bits on heap-(including one bit for deleted)
+   
+    private static final int OFFHEAP_VERSION_MASK = 0x3F_FF_FF;
+
+
+    int getOffHeapVersion(long headerAddress) {
+        return (int) (UnsafeUtils.UNSAFE.getLong(headerAddress) & OFFHEAP_VERSION_MASK);
+    }
+
+    int getDataLength(long headerAddress) {
+        return (int) (UnsafeUtils.UNSAFE.getLong(headerAddress) >>> VERSION_SIZE); //TODO length can be long??
+    }
+    
+    long getOffHeapMetaData(long headerAddress) {
+        return UnsafeUtils.UNSAFE.getLong(headerAddress);
+    }
+
+    private void setDataLength(long headerAddress, int length) {
+        throw new IllegalAccessError();
+    }
+
+    void initHeader(long headerAddress, int dataLength, int version) {
+        long sliceHeader    = (long) (dataLength) << VERSION_SIZE;
```

Comment:
Please take a look on the classes/files `UnionCodec.java` and `ReferenceCodec.java` they are comfortable and well tested. So you don't need to do all those bit-wise operations yourself.